### ISSUE TITLE
How to get attention of SASRec and BERT4Rec??

### ISSUE THREAD

**Issue by gongtigigi** [RUCAIBox/RecBole#1122]
Same as title!

**Comment by Sherry-XLL** [comment]
@gongtigigi Hello, maybe I don't understand your question very well. Both SASRec and BERT4Rec use `TransformerEncoder` to get attention. You can find related code in `recbole/model/layers.py` as follows. 

https://github.com/RUCAIBox/RecBole/blob/96ee7088d9c0f635020d6027265e522eb36a8fd8/recbole/model/layers.py#L338-L551

**Comment by gongtigigi** [comment]
After training, I use `full_sort_predict` to get single prediction.
I want to know which item affected the prediction, maybe by get attention of TransformerEncoder each session

**Comment by Sherry-XLL** [comment]
@Sherry-XLL The output of `TransformerEncoder` is a list consists of all transformer layers' output, or a list only consists of the output of last transformer layer. If you want to get the attention scores, maybe you can output `attention_scores` 
 or `attention_probs` in `recbole/model/layers.py`.

https://github.com/RUCAIBox/RecBole/blob/96ee7088d9c0f635020d6027265e522eb36a8fd8/recbole/model/layers.py#L387-L399

**Comment by gongtigigi** [comment]
Is there no prepared function to get attention?
`attention_scores` and `attention_probs` are `[batch_size, n_heads, MAX_ITEM_LIST_LENGTH, MAX_ITEM_LIST_LENGTH]`, I guess.
What I want to know is influence on prediction.
Then I take `[batch_size, n_heads, MAX_ITEM_LIST_LENGTH, 0]` which correspond to `[CLS]` to see attention. Is this right?

**Comment by Sherry-XLL** [comment]
@gongtigigi There's no prepared function to get attention of each item on the prediction. In fact, in the sequential recommendation model, the first token is usually not `CLS`, while BERT4Rec will add a `mask_token` after the item sequence. If you want to know the influence of each item in `item_sequence` on the prediction, you can use `score = torch.sum(seq_output * self.item_embedding(pos_items), dim=-1)` to do so.

**Comment by gongtigigi** [comment]
It seems `seq_output` mean the output of the last self-attention layer `(b)` on last sequence `t` of the session, `F^(b)_t`. Is this right?
But I do not mean whole item set of all sessions.
Just few items of the current session.
I want to rank the items in the current session with how influenceable to prediction.
But.. SASRec and BERT4Rec have self-attention layer + FCNN in each block.
So, `[batch_size, n_heads, MAX_ITEM_LIST_LENGTH, t]` is not enough to measure influenceable of items..?

**Comment by Sherry-XLL** [comment]
@gongtigigi Sequential recommendation aims to recommend next item that the user will likely interact in a near future, and the next item after the specific `item_seq` is the prediction you mean. In SASRec and BERT4Rec, we want to utilize `TransformerEncoder` to get attention and use the output of last transformer layer in the specific position to make predictions.

`seq_output` gathers the vectors at the specific positions (`item_seq_len - 1`) over a minibatch in the last layer of self-attention's output. The shape of  `seq_output` is `[batch_size, hidden_size]`, which means the hidden embeddings of last item in each item sequence over a minibatch. 

In the current models, inner product between `seq_output` and `item_embedding` can be regarded as the probability distribution of the next item in all items. If our goal is to know the impact of each item in the item sequence on the prediction results, the inner product between `seq_output` and item embedding of `item_seq` can be used as the influence score of each item in the sequence on the prediction result.

Let's take a simple example. For a user who has interacted with 50 items in chronological order, we can use SASRec or BERT4Rec to predict the next item he/she wants to interact with, and get the impact of these 50 items on the prediction results. In this case, `item_seq` is `[i0, i1, ..., i49]`，`item_seq_len` is 50. 

Due to the attention mechanism, each item gets attention of all items. What we want to know is the `i50`,  and we use the attention output of item i49 (`seq_output`) for prediction considering chronological order and position embedding. In BERT4Rec, `reconstruct_test_data`  adds mask token at the last position according to the lengths of `item_seq`, and the output embedding of the mask token acts as `seq_output`. 

If we want to know how each item in `item_seq` influences the prediction, we can do inner product between item embedding of `item_seq` and `seq_output`. That is, `torch.sum(seq_output * self.item_embedding(item_seq), dim=1)`. More specifically, the score of item `i1` on the prediction result is `torch.sum(seq_output * self.item_embedding(i1), dim=1)`. 

In this case, you'd better add one more mask token at the last position like BERT4Rec, and the mask token can be approximately considered as `CLS` token in Transformer, which gets attention of every item in `item_seq`. Finally, inner product between embeddings of 50 items and the output embedding of the token at the last position is the influence score of each item on the prediction with the shape `[50]`. By sorting the scores, we can know which items have a greater impact on the prediction.

**Comment by gongtigigi** [comment]
Thanks for your detailed explanation
Many many thanks!!
